File: CS/E3747CS/FilterHelper.cs
using System;
using System.Collections.Generic;
using System.Text;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.Data;
using System.Data;
using System.ComponentModel;
using DevExpress.Data.Filtering.Helpers;
using System.Collections;

namespace FilterMasterDetailGrid
{
    public class FilterHelper
    {
        GridView _parentView;
        GridView _childView;

        public FilterHelper(GridView parentView, GridView childView)
        {
            _parentView = parentView;
            _parentView.CustomRowFilter += _view_CustomRowFilter;
            _childView = childView;
        }

        void _view_CustomRowFilter(object sender, DevExpress.XtraGrid.Views.Base.RowFilterEventArgs e)
        {
            GridView view = (sender as GridView);
            BaseGridController controller = view.DataController;
            PropertyDescriptorCollection pdc = null;
                IList source = (IList)view.DataSource;
            if (view.DataSource is ITypedList)
            {
                pdc = (view.DataSource as ITypedList).GetItemProperties(null);
            }
            else
            {
                pdc = TypeDescriptor.GetProperties(source.GetType().GetProperty("Item").PropertyType);
            }
            ExpressionEvaluator ev = new ExpressionEvaluator(pdc, controller.FilterCriteria);
            e.Visible = !IsEmptyDetail(e.ListSourceRow, controller) && ev.Fit(source[e.ListSourceRow]);
            e.Handled = true;
        }


        private bool IsEmptyDetail(int listSourceRow, BaseGridController controller)
        {
            IList detail = (IList)controller.GetDetailList(listSourceRow, 0);
            PropertyDescriptorCollection pdc = null;
            if (detail is ITypedList)
            {
                pdc = (detail as ITypedList).GetItemProperties(null);
            }
            else
            {
                pdc = TypeDescriptor.GetProperties(detail.GetType().GetProperty("Item").PropertyType);
            }
            ExpressionEvaluator detailEvaluator = new ExpressionEvaluator(pdc, _childView.DataController.FilterCriteria);
            foreach (object o in detail)
            {
                if (detailEvaluator.Fit(o))
                    return false;
            }
            return true;
        }
    }
}

File: VB/E3747VB/FilterHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.Data
Imports System.Data
Imports System.ComponentModel
Imports DevExpress.Data.Filtering.Helpers
Imports System.Collections

Namespace FilterMasterDetailGrid
	Public Class FilterHelper
		Private _parentView As GridView
		Private _childView As GridView

		Public Sub New(ByVal parentView As GridView, ByVal childView As GridView)
			_parentView = parentView
			AddHandler _parentView.CustomRowFilter, AddressOf _view_CustomRowFilter
			_childView = childView
		End Sub

		Private Sub _view_CustomRowFilter(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Base.RowFilterEventArgs)
			Dim view As GridView = (TryCast(sender, GridView))
			Dim controller As BaseGridController = view.DataController
			Dim pdc As PropertyDescriptorCollection = Nothing
				Dim source As IList = CType(view.DataSource, IList)
			If TypeOf view.DataSource Is ITypedList Then
				pdc = (TryCast(view.DataSource, ITypedList)).GetItemProperties(Nothing)
			Else
				pdc = TypeDescriptor.GetProperties(CType(source, Object).GetType().GetProperty("Item").PropertyType)
			End If
			Dim ev As New ExpressionEvaluator(pdc, controller.FilterCriteria)
			e.Visible = (Not IsEmptyDetail(e.ListSourceRow, controller)) AndAlso ev.Fit(source(e.ListSourceRow))
			e.Handled = True
		End Sub


		Private Function IsEmptyDetail(ByVal listSourceRow As Integer, ByVal controller As BaseGridController) As Boolean
			Dim detail As IList = CType(controller.GetDetailList(listSourceRow, 0), IList)
			Dim pdc As PropertyDescriptorCollection = Nothing
			If TypeOf detail Is ITypedList Then
				pdc = (TryCast(detail, ITypedList)).GetItemProperties(Nothing)
			Else
				pdc = TypeDescriptor.GetProperties(CType(detail, Object).GetType().GetProperty("Item").PropertyType)
			End If
			Dim detailEvaluator As New ExpressionEvaluator(pdc, _childView.DataController.FilterCriteria)
			For Each o As Object In detail
				If detailEvaluator.Fit(o) Then
					Return False
				End If
			Next o
			Return True
		End Function
	End Class
End Namespace
